### PR TITLE
Support new instructions in HiPE

### PR DESCRIPTION
**Pull request by michael-swan**
This is almost complete. What remains:

1. Adding support for 32-bit bs_start_match3. Currently it just emits code equivalent to the 64-bit emulator code.
2. Remove any trace of bs_start_match2 throughout the HiPE compiler, barring objection from maintainers.
3. Refactor my bs_start_match3 handling code. I copied a number of existing functions and modified them slightly to suit my immediate interests. This is messy and needs to be replaced. What this refactor looks like depends on whether or not we remove bs_start_match2 support from HiPE.
4. Thoroughly test this feature to ensure this generated code is not going to cause issues.

I will try to get 32-bit support up tomorrow. In the meantime, let me know what the feeling is on bs_start_match2 getting tossed from HiPE. It is of little cost to keep it, it just affects how I refactor things a little.

### CODE REVIEW

**Comment by jhogberg** [comment]
Thanks for the PR, your help is much appreciated!

Feel free to skip `bs_start_match2`, the compiler only emits it when given undocumented options intended for use in testing, and we may remove it altogether in a few years.

**Comment by michael-swan** [comment]
@mikpe Thanks for taking a look. I will try that out and see what sort of results we get, but I also wanted to go through and trace some examples using these features in both BEAM and HiPE and try to hit every branch, doing a side-by-side comparison of the output (e.g. match state).

The one thing I know to be wrong right now is 32-bit bs_get_position and bs_set_position. That will be trivial to add support for.

@jhogberg I am looking at your 32-bit `bs_set_position` emulator implementation and it looks like you run the risk of overflowing your `ms->mb.offset` value in the big integer case. I guess it could be the case that the compiler somehow prevents this kind of usage of the bs_set_position instruction or there is some invariant in that UInt in big ints. Not too familiar with this code. If this were to overflow this could present the risk of information disclosure because now your offset from the beginning of your binary is negative, although downstream instructions may perform checks to prevent that, I haven't looked any further.

**Comment by michael-swan** [comment]
Alright, I am pretty sure this is done, but I would really prefer to first go through the tediousness that is actually testing this thing before calling it "done." I am thinking I am going to run this thing through GDB with a few test programs that is sufficient to get full coverage of these instructions in both the emulator and this HiPE code. I wish there was a way to insert a "print this address" HiPE RTL pseudo-instructions so I could easily find where each of these basic blocks lies once finally loaded. It has so far been a bit of a pain to drink from the firehose of machine instructions I am not at all concerned with. I'll figure something out.

File: lib/hipe/rtl/hipe_rtl_binary_match.erl [hunk 1]
```diff
@@ -103,6 +148,68 @@ gen_rtl({{bs_start_match, ok_matchstate}, _Max}, [], [Binary],
...
+				  hipe_rtl:label_name(SmallLbl),
+				  hipe_rtl:label_name(BigLbl),
+				  0.99),
+       SmallLbl,
+       hipe_tagscheme:realuntag_fixnum(Tmp, Pos),
+       hipe_rtl:mk_goto(hipe_rtl:label_name(RestLbl)),
+       BigLbl,
+       unsafe_get_one_word_pos_bignum(Tmp, Pos),
+       RestLbl,
+       hipe_rtl:mk_alu(Tmp, Tmp, add, SaveOffset)
```
Comment (llelf) [inline lib/hipe/rtl/hipe_rtl_binary_match.erl]: @michael-swan  Mis-merge?
```erl
 hipe_tagscheme:unsafe_get_one_word_pos_bignum(Tmp, Pos), % module
 RestLbl,
 hipe_rtl:mk_alu(Tmp, Tmp, add, SaveOffset), % comma
```

Comment (michael-swan) [reply lib/hipe/rtl/hipe_rtl_binary_match.erl]: Yes. Thank you.

**Comment by michael-swan** [comment]
I have been extra super busy the past two months, but I plan to perform a reasonable amount of testing on this two weekend from now when things normalize again! Space is allocated on my calendar, so it will definitely happen barring other stuff coming up!

**Comment by mikpe** [comment]
Given that HiPE has been removed from master, I think this should be declined.

**Comment by bjorng** [comment]
Thanks for your work on HiPE. Since HiPE will be removed in OTP 24, we will close this pull request.

**Comment by michael-swan** [comment]
Yeah no worries. Sorry I never got back to double checking any of this anyways. Glad to hear y'all are making progress. HiPE was such a mess.

**Pull request closed without merging**